src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs

```diff
@@ -208,6 +208,104 @@ internal sealed override IEnumerable<Cci.INestedTypeDefinition> GetSynthesizedNe
             return null;
         }
 
+        public sealed override ImmutableArray<(Cci.ITypeDefinition, ImmutableArray<Cci.DebugSourceDocument>)> GetTypeToDebugDocumentMap(EmitContext context)
```

Comment:
>ImmutableArray<(Cci.ITypeDefinition, ImmutableArray<Cci.DebugSourceDocument>)> [](http://example.com/codeflow?start=31&length=78)

Could we reduce the allocations by having the caller pass in a callback that gets called once per type?
```csharp
public sealed override void GetTypeToDebugDocuments(
    EmitContext context,
    Action<MetadataBuilder builder, Cci.ITypeDefinition, ArrayBuilder<Cci.DebugSourceDocument>> callback);
```
Alternatively, perhaps return an `IEnumerable<(Cci.ITypeDefinition, ImmutableArray<Cci.DebugSourceDocument>)>` rather than `ImmutableArray<(..., ...)>` and use `yield return` here, to match other `PEModuleBuilder` methods and to avoid allocating a single array.

Reply:
I don't know why we need to reduce allocations here and make this that complicated. Emit is generally not optimized to minimize allocations.

Reply:
> I don't know why we need to reduce allocations here and make this that complicated.

Is there a reason to return `ImmutableArray<(..., ...)>` rather than `IEnumerable<(..., ...)>` (with `yield return` instead of a large array)? `IEnumerable<(..., ...)>` seems consistent with other methods here and no more complicated.

Reply:
IEnumerable make sense. BTW. I don't expect the result to be large.